solutions/gold/cf-468B.mdx

```diff
@@ -0,0 +1,100 @@
+---
+id: cf-468B
+source: CF
+title: Two Sets
+author: Chongtian Ma
+---
+
+## Explanation
+
+We can treat each set as a graph with several connected components. For each $p_i$, if $a - p_i$ exists, then $p_i$ and $a - p_i$ are in the same connected component inside graph $A$, and similarly for $B$. Each number in the connected component must belong to one set. Therefore, our condradiction comes from if there is a number $x$ in the component such that only $a - x$ exists, and another $y$ in the component such that only $b - y$ exists, or vice cersa.
+
+ We can check each component for the contradiction using DSU. The status of the whole component will be the intersection of what graph each number in the component can reside in.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+struct DSU {
```

Comment:
code snip + use the one from the module for consistency

Reply:
do i have to

Reply:
you can always have ryan merge

Reply:
huh, i didn't realize that the implementation for the concept switched over to one with classes...

i feel like that can be pretty unreadable too :p

probably shouldn't use your own, but this was the past implementation 

```cpp
struct DSU {
	vector<int> e;
	DSU(int N) { e = vector<int>(N, -1); }

	// get representive component (uses path compression)
	int get(int x) { return e[x] < 0 ? x : e[x] = get(e[x]); }
	
	bool same_set(int a, int b) { return get(a) == get(b); }
	
	int size(int x) { return -e[get(x)]; }
	
	bool unite(int x, int y) {  // union by size
		x = get(x), y = get(y);
		if (x == y) return false;
		if (e[x] > e[y]) swap(x, y);
		e[x] += e[y]; e[y] = x;
		return true;
	}
};
```
if you'd like to use that one, im more than happy to merge :D